### PR TITLE
[staking] Candidate Register without Staking

### PR DESCRIPTION
**Pull request by envestcc**
# Description

We expanded the action `CandidateRegister` to support registration without staking. However, in the absence of self-staking, they will not be selected as consensus nodes.

## Type of change
Please delete options that are not relevant.
- [x] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/staking/validations.go [hunk 1]
```diff
@@ -66,7 +70,11 @@ func (p *Protocol) validateCandidateRegister(ctx context.Context, act *action.Ca
 	}
 
 	if act.Amount().Cmp(p.config.RegistrationConsts.MinSelfStake) < 0 {
-		return errors.Wrap(action.ErrInvalidAmount, "self staking amount is not valid")
+		featureCtx := protocol.MustGetFeatureCtx(ctx)
+		if featureCtx.CandidateRegisterMustWithStake || act.Amount().Cmp(big.NewInt(0)) != 0 {
```
Comment (CoderZhi) [inline action/protocol/staking/validations.go:74]: `if featureCtx.CandidateRegisterMustWithStake` always return error?

Comment (envestcc) [reply action/protocol/staking/validations.go:74]: yes, only `featureCtx.CandidateRegisterMustWithStake==false && act.Amount()==0` is valid if `act.Amount() < MinSelfStake`

File: action/protocol/staking/handlers.go [hunk 2]
```diff
@@ -663,19 +663,33 @@ func (p *Protocol) handleCandidateRegister(ctx context.Context, act *action.Cand
...
 	}
 
-	bucket := NewVoteBucket(owner, owner, act.Amount(), act.Duration(), blkCtx.BlockTimeStamp, act.AutoStake())
-	bucketIdx, err := csm.putBucketAndIndex(bucket)
-	if err != nil {
-		return log, nil, err
+	// register with self-stake
+	bucketIdx := uint64(candidateNoSelfStakeBucketIndex)
+	var bucketLogs []*action.TransactionLog
+	votes := big.NewInt(0)
```
Comment (dustinxie) [inline action/protocol/staking/handlers.go:669]: use `var ()` to put 3 together

File: action/protocol/staking/candidate_center.go [hunk 3]
```diff
@@ -529,7 +532,7 @@ func (cb *candBase) collision(d *Candidate) (address.Address, address.Address, a
 		oper = c.Owner
 	}
 
-	if c, hit := cb.selfStkBucketMap[d.SelfStakeBucketIdx]; hit && !address.Equal(c.Owner, d.Owner) {
+	if c, hit := cb.selfStkBucketMap[d.SelfStakeBucketIdx]; hit && d.SelfStakeBucketIdx != candidateNoSelfStakeBucketIndex && !address.Equal(c.Owner, d.Owner) {
```
Comment (dustinxie) [inline action/protocol/staking/candidate_center.go:535]: change is this file should use what's in PR4061, right?

File: action/protocol/staking/handlers.go [hunk 4]
```diff
@@ -663,19 +663,33 @@ func (p *Protocol) handleCandidateRegister(ctx context.Context, act *action.Cand
...
+	if act.Amount().Sign() > 0 {
+		bucket := NewVoteBucket(owner, owner, act.Amount(), act.Duration(), blkCtx.BlockTimeStamp, act.AutoStake())
+		bktIdx, err := csm.putBucketAndIndex(bucket)
+		if err != nil {
+			return log, nil, err
+		}
+		bucketIdx = bktIdx
```
Comment (dustinxie) [inline action/protocol/staking/handlers.go:676]: L672: `bucketIdx, err = csm.xxx`, no need to have a new var `bktIdx` here

File: action/protocol/staking/handlers.go [hunk 5]
```diff
@@ -688,49 +702,43 @@ func (p *Protocol) handleCandidateRegister(ctx context.Context, act *action.Cand
...
-			Amount:    act.Amount(),
-		},
-		{
-			Type:      iotextypes.TransactionLogType_CANDIDATE_REGISTRATION_FEE,
-			Sender:    actCtx.Caller.String(),
-			Recipient: address.RewardingPoolAddr,
-			Amount:    registrationFee,
-		},
-	}, nil
+	txLogs := append(bucketLogs, &action.TransactionLog{
```
Comment (dustinxie) [inline action/protocol/staking/handlers.go:735]: no need to define a new `txLogs`? can use existing `bucketLogs`
i think rename `bucketLogs` to `txLogs` to make the name more clear

File: action/protocol/staking/handlers.go [hunk 6]
```diff
@@ -663,19 +663,33 @@ func (p *Protocol) handleCandidateRegister(ctx context.Context, act *action.Cand
...
+		}
+		bucketIdx = bktIdx
+		bucketLogs = append(bucketLogs, &action.TransactionLog{
+			Type:      iotextypes.TransactionLogType_CANDIDATE_SELF_STAKE,
+			Sender:    actCtx.Caller.String(),
+			Recipient: address.StakingBucketPoolAddr,
+			Amount:    act.Amount(),
+		})
+		votes = p.calculateVoteWeight(bucket, true)
+		log.AddTopics(byteutil.Uint64ToBytesBigEndian(bucketIdx), owner.Bytes())
```
Comment (dustinxie) [inline action/protocol/staking/handlers.go:684]: this needs to be outside of `if{}` -- registration without self-stake should also create a topic in the log

File: action/protocol/staking/candidate_center.go (hunk 3 above)
Comment (envestcc) [reply action/protocol/staking/candidate_center.go:535]: yes, it's part of PR4061

**Review by CoderZhi — APPROVED** [review]

File: action/protocol/staking/validations.go [hunk 7]
```diff
@@ -66,7 +70,11 @@ func (p *Protocol) validateCandidateRegister(ctx context.Context, act *action.Ca
...
+			log.L().Warn("Candidate register amount is less than the minimum requirement", zap.Bool("feature", featureCtx.CandidateRegisterMustWithStake), zap.String("amount", act.Amount().String()))
```
Comment (Liuhaai) [inline action/protocol/staking/validations.go:75]: i think the error has been logged upstream

File: action/protocol/staking/validations.go (hunk 1 above)
Comment (Liuhaai) [reply action/protocol/staking/validations.go:74]: !featureCtx.CandidateRegisterMustWithStake  && act.Amount()==0 is easier to read than featureCtx.CandidateRegisterMustWithStake || act.Amount().Cmp(big.NewInt(0)) != 0

File: action/protocol/staking/handlers.go [hunk 8]
```diff
@@ -663,10 +663,26 @@ func (p *Protocol) handleCandidateRegister(ctx context.Context, act *action.Cand
...
-	if err != nil {
-		return log, nil, err
+	// register with self-stake
+	var (
+		bucketIdx = uint64(candidateNoSelfStakeBucketIndex)
+		txLogs    []*action.TransactionLog
+		votes     = big.NewInt(0)
+		err       error
+	)
+	if act.Amount().Sign() > 0 {
```
Comment (Liuhaai) [inline action/protocol/staking/handlers.go:673]: no hardfork here?

Comment (envestcc) [reply action/protocol/staking/handlers.go:673]: Before the hard fork, the `act.Amount` must be greater than 0, behaves the same as in the old version.

Comment (millken) [reply action/protocol/staking/handlers.go:712]: `if act.Amount().Sign() <= 0 {` should be return error?

File: action/protocol/staking/handlers.go [hunk 9]
```diff
@@ -702,10 +702,26 @@ func (p *Protocol) handleCandidateRegister(ctx context.Context, act *action.Cand
 		}
 	}
 
-	bucket := NewVoteBucket(owner, owner, act.Amount(), act.Duration(), blkCtx.BlockTimeStamp, act.AutoStake())
-	bucketIdx, err := csm.putBucketAndIndex(bucket)
-	if err != nil {
-		return log, nil, err
+	// register with self-stake
```
Comment (dustinxie) [inline action/protocol/staking/handlers.go:705]: move to L713

File: action/protocol/staking/handlers.go [hunk 10]
```diff
@@ -702,10 +702,26 @@ func (p *Protocol) handleCandidateRegister(ctx context.Context, act *action.Cand
...
+			return log, nil, err
+		}
+		txLogs = append(txLogs, &action.TransactionLog{
+			Type:      iotextypes.TransactionLogType_CANDIDATE_SELF_STAKE,
+			Sender:    actCtx.Caller.String(),
+			Recipient: address.StakingBucketPoolAddr,
+			Amount:    act.Amount(),
+		})
+		votes = p.calculateVoteWeight(bucket, true)
 	}
```
Comment (dustinxie) [inline action/protocol/staking/handlers.go:725]: think adding an else make the code more explicit
```
else {
    // register w/o self-stake, waiting to be endorsed
   bucketIdx = uint64(candidateNoSelfStakeBucketIndex)
   votes  = big.NewInt(0)
}
```

File: action/protocol/staking/handlers.go [hunk 11]
```diff
@@ -702,10 +702,26 @@ func (p *Protocol) handleCandidateRegister(ctx context.Context, act *action.Cand
...
-	bucket := NewVoteBucket(owner, owner, act.Amount(), act.Duration(), blkCtx.BlockTimeStamp, act.AutoStake())
-	bucketIdx, err := csm.putBucketAndIndex(bucket)
-	if err != nil {
-		return log, nil, err
+	// register with self-stake
+	var (
+		bucketIdx = uint64(candidateNoSelfStakeBucketIndex)
+		txLogs    []*action.TransactionLog
+		votes     = big.NewInt(0)
+		err       error
```
Comment (dustinxie) [inline action/protocol/staking/handlers.go:710]: 1. withSelfStake = act.Amount().Sign() > 0
2. `err error` not ncessary?

File: action/protocol/staking/handlers.go [hunk 12]
```diff
@@ -702,10 +702,26 @@ func (p *Protocol) handleCandidateRegister(ctx context.Context, act *action.Cand
...
+	)
+	if act.Amount().Sign() > 0 {
```
Comment (dustinxie) [inline action/protocol/staking/handlers.go:712]: if withSelfStake {

File: action/protocol/staking/handlers.go [hunk 13]
```diff
@@ -727,49 +743,43 @@ func (p *Protocol) handleCandidateRegister(ctx context.Context, act *action.Cand
 		csm.DirtyView().candCenter.base.recordOwner(c)
 	}
 
-	// update bucket pool
-	if err := csm.DebitBucketPool(act.Amount(), true); err != nil {
-		return log, nil, &handleError{
-			err:           errors.Wrapf(err, "failed to update staking bucket pool %s", err.Error()),
-			failureStatus: iotextypes.ReceiptStatus_ErrWriteAccount,
+	if act.Amount().Sign() > 0 {
```
Comment (dustinxie) [inline action/protocol/staking/handlers.go:746]: if withSelfStake {

File: action/protocol/staking/handlers.go (hunk 8 above)
Comment (envestcc) [reply action/protocol/staking/handlers.go:712]: already checked at https://github.com/iotexproject/iotex-core/pull/4059/files#diff-49d3555713cf037eb3ff80b28e13bd7329cc276b8fb0d60e4798b055a7e4aad8R70

**Pull request merged**